### PR TITLE
using -1e+20 as wells' reference bhp depth

### PR DESCRIPTION
**Pull request by GitPaean**
for wells with defaulted depth while do not have active connections

### CODE REVIEW

**Comment by GitPaean** [comment]
jenkins build this please

**Comment by GitPaean** [comment]
I expect this should not change any current behavoir, except trying to fix https://github.com/OPM/opm-common/issues/2442 . 

But it does not update the reference depth when filtering off some connections, but I guess we do not have case for it yet.  Current master branch does not handle it either.

**Comment by GitPaean** [comment]
jenkins build this please

File: src/opm/parser/eclipse/EclipseState/Schedule/Well/Well.cpp [hunk 1]
```diff
@@ -812,8 +812,10 @@ bool Well::getAllowCrossFlow() const {
 }
 
 double Well::getRefDepth() const {
-    if (!this->ref_depth.has_value())
-        throw std::logic_error(fmt::format("Well: {} - tried to access not initialized well reference depth", this->name()));
+    if (!this->ref_depth)
```
Comment (bska) [inline src/opm/parser/eclipse/EclipseState/Schedule/Well/Well.cpp:815]: Please keep the `has_value` call here.

File: src/opm/parser/eclipse/EclipseState/Schedule/Well/Well.cpp [hunk 2]
```diff
@@ -812,8 +812,10 @@ bool Well::getAllowCrossFlow() const {
...
+        // for wells with default bhp reference depth while no connections defined,
+        // using this specific value following commercial simulation reference result
+        return -1.e+20;
```
Comment (bska) [inline src/opm/parser/eclipse/EclipseState/Schedule/Well/Well.cpp:818]: The callers of this function expects SI values and there are at least some callers&mdash;notably in the output/reporting code&mdash;that will convert the value to output units for display/output purposes.  A bare `-1.0e+20` will coincidentally do the right thing for METRIC and PVT-M units but will be wrong in LAB and FIELD units. You will also have to look at the callers and add special casing for negative return values as needed to ensure we always output the exact value `-1.0e+20` regardless of unit system.

**Review by bska — COMMENTED** [review]
I this may be a step in the right direction to resolve #2442, but it's not quite sufficient in its current form.

File: src/opm/parser/eclipse/EclipseState/Schedule/Well/Well.cpp (hunk 2 above)
Comment (GitPaean) [reply src/opm/parser/eclipse/EclipseState/Schedule/Well/Well.cpp:818]: Not sure what behavoir we should expect under different unit system. Created a test case https://github.com/OPM/opm-common/issues/2442 , hopefully we can figure it out with a test case.

File: src/opm/parser/eclipse/EclipseState/Schedule/Well/Well.cpp (hunk 1 above)
Comment (GitPaean) [reply src/opm/parser/eclipse/EclipseState/Schedule/Well/Well.cpp:815]: Sure.

**Comment by GitPaean** [comment]
jenkins build this please

File: src/opm/parser/eclipse/EclipseState/Schedule/Well/Well.cpp (hunk 2 above)
Comment (bska) [reply src/opm/parser/eclipse/EclipseState/Schedule/Well/Well.cpp:818]: > Not sure what behaviour we should expect under different unit system.

The constant `-1.0e+20` is ECLIPSE's typical sentinel value for "not defined" quantities.  In other words, if ECLIPSE outputs `-1.0e+20` metres in METRIC units, then we should output `-1.0e+20` feet in FIELD units and `-1.0e+20` centimetre in LAB units.  In other words, the constant `-1.0e+20` must **not** be subject to unit conversion, meaning we must insert special case code to intercept (large) negative values and pass those directly on to the output destination.

**Comment by joakim-hove** [comment]
Maybe it is better to just return an optional; then the numerical value `-1.0e+20` can just be entered right at the point of output - irrespective of unit system used.

**Comment by bska** [comment]
> Maybe it is better to just return an optional;

That would solve the problem for the output code.  Whether or not it is sufficient for the parallel communication/distribution of the `Schedule` object depends on our support for communicating `std::optional<>` objects.

**Comment by GitPaean** [comment]
> The constant -1.0e+20 is ECLIPSE's typical sentinel value for "not defined" quantities. In other words, if ECLIPSE outputs -1.0e+20 metres in METRIC units, then we should output -1.0e+20 feet in FIELD units and -1.0e+20 centimetre in LAB units. In other words, the constant -1.0e+20 must not be subject to unit conversion, meaning we must insert special case code to intercept (large) negative values and pass those directly on to the output destination.

And also we need to consider output and restart input(do we?), it is better to fix through the I/O code?

> Maybe it is better to just return an optional;

Something like that. Maybe instead of using optional, we need a extra flag in the well code to indicate whether the bhp is defined (or a function combine whether defaulted bhp from WELSPECS and whether there is active connections to indicate whether we should access the bhp value.)


It looks like this fix might go farther out of the class `Well` to I/O part, maybe better @bska or @joakim-hove fix it. I am not very available the rest of the week.  Feel free to close this PR.

**Comment by bska** [comment]
I'm going to close this now.  We have solved the underlying issue in a different way.

**Pull request closed without merging**